pkg/util/hostname.go

```diff
@@ -121,54 +155,192 @@ func saveHostnameData(cacheHostnameKey string, hostname string, provider string)
 	return hostnameData
 }
 
+// ResolveSourcesWithState will merge two HostnameMap sources, providing the expected
+// resolved output.
+func ResolveSourcesWithState(newSources, stateSources HostnameMap) (HostnameMap, bool) {
+	resolvedSources := HostnameMap{}
+	stateChange := false
+
+	if stateSources == nil {
+		return newSources, true
+	}
+
+	for _, stage := range resolutionPipeline {
+		log.Debug("Getting hostname collected by: %s", stage.provider)
+
+		newH, newOk := newSources[stage.provider]
+		stateH, stateOk := stateSources[stage.provider]
+
+		if !newOk && !stateOk {
+			// missing source
+			continue
+		}
+
+		h := newH
+		if stage.reliable {
+			if newOk {
+				if newH != stateH {
+					stateChange = true
+				}
+			} else {
+				if stateOk {
+					stateChange = true
+					log.Warnf("Reliable source %s no longer configured", stage.provider)
+				}
+
+				// reliable source did not resolve thus it does not apply
+				continue
+			}
+		} else {
+			if newOk {
+				if newH != stateH {
+					stateChange = true
+				}
+			} else {
+				if stateOk {
+					// here we should always use stored state - an unreliable source failed
+					h = stateH
+					log.Info("an unreliable source %s did not resolve a hostname as expected, using stored state: %s",
+						stage.provider, stateH)
+				} else {
+					// resolution stage did not resolve as expected
+					continue
+				}
+			}
+		}
```

Comment:
This looks about right, we only want to `continue` if `stage.reliable` is true though. Otherwise we want to go through that logic, but continue through the rest of the loop normally.